client/remote.go

```diff
@@ -369,38 +370,43 @@ func (g *Group) Dial(c *Client) (conn *Conn, err error) {
 
 	// loop through all remote servers for performance measurement
 	// in a separate goroutine
-	go func() {
-		time.Sleep(100 * time.Microsecond)
-		g.Lock()
-		for _, i := range g.remotes {
-			conn, err := i.Dial(c)
-			if err != nil {
-				i.latency.Reset()
-				i.errorCount++
-				log.Infof("ping failed: %v", err)
-				continue
-			}
+	if time.Since(g.lastPingAll) > time.Second {
+		g.lastPingAll = time.Now()
+		go g.pingAll(c)
+	}
 
-			start := time.Now()
-			err = conn.Ping(nil)
-			duration := time.Since(start)
+	return conn, nil
+}
 
-			if err != nil {
-				defer conn.Close()
-				i.latency.Reset()
-				i.errorCount++
-				log.Infof("ping failed: %v", err)
-			} else {
-				log.Debug(conn.addr, " ping duration:", duration)
-				i.latency.Update(duration)
-			}
+func (g *Group) pingAll(c *Client) {
+	time.Sleep(100 * time.Microsecond)
+	g.Lock()
```

Comment:
I actually want to rewrite this part as well. Do you mind reviewing my changes soonish?
The "pingAll" mechanism shouldn't lock the group in practical uses. Because a group can have hundreds of remotes, pinging each of them sequentially takes a lot of time, factoring in timeout/error etc. So I decide to make latency measurement less aggressive and more opportunistic.